File: pulse-kotlin/src/commonTest/kotlin/com/fluxchess/pulse/kotlin/fen/FenTest.kt
/*
 * Copyright 2013-2025 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.fen

import com.fluxchess.pulse.kotlin.engine.A1
import com.fluxchess.pulse.kotlin.engine.A8
import com.fluxchess.pulse.kotlin.engine.B1
import com.fluxchess.pulse.kotlin.engine.B8
import com.fluxchess.pulse.kotlin.engine.BLACK_BISHOP
import com.fluxchess.pulse.kotlin.engine.BLACK_KING
import com.fluxchess.pulse.kotlin.engine.BLACK_KINGSIDE
import com.fluxchess.pulse.kotlin.engine.BLACK_KNIGHT
import com.fluxchess.pulse.kotlin.engine.BLACK_PAWN
import com.fluxchess.pulse.kotlin.engine.BLACK_QUEEN
import com.fluxchess.pulse.kotlin.engine.BLACK_QUEENSIDE
import com.fluxchess.pulse.kotlin.engine.BLACK_ROOK
import com.fluxchess.pulse.kotlin.engine.C1
import com.fluxchess.pulse.kotlin.engine.C8
import com.fluxchess.pulse.kotlin.engine.D1
import com.fluxchess.pulse.kotlin.engine.D8
import com.fluxchess.pulse.kotlin.engine.E1
import com.fluxchess.pulse.kotlin.engine.E8
import com.fluxchess.pulse.kotlin.engine.F1
import com.fluxchess.pulse.kotlin.engine.F8
import com.fluxchess.pulse.kotlin.engine.G1
import com.fluxchess.pulse.kotlin.engine.G8
import com.fluxchess.pulse.kotlin.engine.H1
import com.fluxchess.pulse.kotlin.engine.H8
import com.fluxchess.pulse.kotlin.engine.Position
import com.fluxchess.pulse.kotlin.engine.RANK_2
import com.fluxchess.pulse.kotlin.engine.RANK_7
import com.fluxchess.pulse.kotlin.engine.WHITE
import com.fluxchess.pulse.kotlin.engine.WHITE_BISHOP
import com.fluxchess.pulse.kotlin.engine.WHITE_KING
import com.fluxchess.pulse.kotlin.engine.WHITE_KINGSIDE
import com.fluxchess.pulse.kotlin.engine.WHITE_KNIGHT
import com.fluxchess.pulse.kotlin.engine.WHITE_PAWN
import com.fluxchess.pulse.kotlin.engine.WHITE_QUEEN
import com.fluxchess.pulse.kotlin.engine.WHITE_QUEENSIDE
import com.fluxchess.pulse.kotlin.engine.WHITE_ROOK
import com.fluxchess.pulse.kotlin.engine.files
import com.fluxchess.pulse.kotlin.engine.squareOf
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class FenTest {
	@Test
	fun `Valid FEN should return a valid position`() {
		val position = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1".toPosition()
		assertEquals(standardPosition(), position)
	}

	@Test
	fun `Invalid FEN should return an error`() {
		val exception = assertFailsWith<FenException> { "invalid FEN".toPosition() }
		assertEquals("Invalid FEN: invalid FEN", exception.message)
	}

	@Test
	fun `Invalid ranks should return an error`() {
		val exception = assertFailsWith<FenException> { "invalid-ranks w KQkq - 0 1".toPosition() }
		assertEquals("Invalid board: invalid-ranks", exception.message)
	}

	@Test
	fun `Invalid rank with too many squares should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr1/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1".toPosition() }
		assertEquals("Invalid rank: rnbqkbnr1", exception.message)
	}

	@Test
	fun `Invalid rank with too few squares should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbn/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1".toPosition() }
		assertEquals("Invalid rank: rnbqkbn", exception.message)
	}

	@Test
	fun `Invalid piece type should return an error`() {
		val exception = assertFailsWith<FenException> { "xnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1".toPosition() }
		assertEquals("Invalid rank: xnbqkbnr", exception.message)
	}

	@Test
	fun `Invalid number of empty squares should return an error`() {
		val exception = assertFailsWith<FenException> { "9/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1".toPosition() }
		assertEquals("Invalid rank: 9", exception.message)
	}

	@Test
	fun `Invalid active color should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR x KQkq - 0 1".toPosition() }
		assertEquals("Invalid active color: x", exception.message)
	}

	@Test
	fun `Invalid castling rights with too many castlings should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkqK - 0 1".toPosition() }
		assertEquals("Invalid castling rights: KQkqK", exception.message)
	}

	@Test
	fun `Invalid castling rights should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w XQkq - 0 1".toPosition() }
		assertEquals("Invalid castling type: X", exception.message)
	}

	@Test
	fun `Invalid en passant square length should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e31 0 1".toPosition() }
		assertEquals("Invalid en passant square: e31", exception.message)
	}

	@Test
	fun `Invalid en passant square with non-existing file should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq x3 0 1".toPosition() }
		assertEquals("Invalid file: x", exception.message)
	}

	@Test
	fun `Invalid en passant square with non-existing rank should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e9 0 1".toPosition() }
		assertEquals("Invalid rank: 9", exception.message)
	}

	@Test
	fun `Invalid en passant square should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e2 0 1".toPosition() }
		assertEquals("Invalid en passant square: e2", exception.message)
	}

	@Test
	fun `Invalid halfmove clock should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - x 1".toPosition() }
		assertEquals("Invalid halfmove clock: x", exception.message)
	}

	@Test
	fun `Invalid fullmove number should return an error`() {
		val exception = assertFailsWith<FenException> { "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 x".toPosition() }
		assertEquals("Invalid fullmove number: x", exception.message)
	}
}

private fun standardPosition(): Position {
	val position = Position()
	position.activeColor = WHITE
	position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
	position.halfmoveNumber = 2
	position.put(WHITE_ROOK, A1)
	position.put(WHITE_KNIGHT, B1)
	position.put(WHITE_BISHOP, C1)
	position.put(WHITE_QUEEN, D1)
	position.put(WHITE_KING, E1)
	position.put(WHITE_BISHOP, F1)
	position.put(WHITE_KNIGHT, G1)
	position.put(WHITE_ROOK, H1)
	position.put(BLACK_ROOK, A8)
	position.put(BLACK_KNIGHT, B8)
	position.put(BLACK_BISHOP, C8)
	position.put(BLACK_QUEEN, D8)
	position.put(BLACK_KING, E8)
	position.put(BLACK_BISHOP, F8)
	position.put(BLACK_KNIGHT, G8)
	position.put(BLACK_ROOK, H8)
	for (file in files) {
		position.put(WHITE_PAWN, squareOf(file, RANK_2))
		position.put(BLACK_PAWN, squareOf(file, RANK_7))
	}
	return position
}

File: pulse-go/internal/pulse/fen/fen_test.go
/*
 * Copyright 2013-2025 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package fen

import (
	"errors"
	"reflect"
	"testing"

	"github.com/fluxroot/pulse/internal/pulse/engine"
)

func TestFEN_ToPosition(t *testing.T) {
	tests := []struct {
		name    string
		fen     string
		want    *engine.Position
		wantErr error
	}{
		{
			name:    "Valid FEN should return a valid position",
			fen:     "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
			want:    standardPosition(),
			wantErr: nil,
		},
		{
			name:    "Invalid FEN should return an error",
			fen:     "invalid FEN",
			want:    nil,
			wantErr: errors.New("invalid FEN: invalid FEN"),
		},
		{
			name:    "Invalid ranks should return an error",
			fen:     "invalid-ranks w KQkq - 0 1",
			want:    nil,
			wantErr: errors.New("invalid board: invalid-ranks"),
		},
		{
			name:    "Invalid rank with too many squares should return an error",
			fen:     "rnbqkbnr1/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
			want:    nil,
			wantErr: errors.New("invalid rank: rnbqkbnr1"),
		},
		{
			name:    "Invalid rank with too few squares should return an error",
			fen:     "rnbqkbn/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
			want:    nil,
			wantErr: errors.New("invalid rank: rnbqkbn"),
		},
		{
			name:    "Invalid piece type should return an error",
			fen:     "xnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
			want:    nil,
			wantErr: errors.New("invalid rank: xnbqkbnr"),
		},
		{
			name:    "Invalid number of empty squares should return an error",
			fen:     "9/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
			want:    nil,
			wantErr: errors.New("invalid rank: 9"),
		},
		{
			name:    "Invalid active color should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR x KQkq - 0 1",
			want:    nil,
			wantErr: errors.New("invalid active color: x"),
		},
		{
			name:    "Invalid castling rights with too many castlings should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkqK - 0 1",
			want:    nil,
			wantErr: errors.New("invalid castling rights: KQkqK"),
		},
		{
			name:    "Invalid castling rights should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w XQkq - 0 1",
			want:    nil,
			wantErr: errors.New("could not map to castling: invalid castling type: X"),
		},
		{
			name:    "Invalid en passant square length should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e31 0 1",
			want:    nil,
			wantErr: errors.New("invalid en passant square: e31"),
		},
		{
			name:    "Invalid en passant square with non-existing file should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq x3 0 1",
			want:    nil,
			wantErr: errors.New("invalid file: x"),
		},
		{
			name:    "Invalid en passant square with non-existing rank should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e9 0 1",
			want:    nil,
			wantErr: errors.New("invalid rank: 9"),
		},
		{
			name:    "Invalid en passant square should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e2 0 1",
			want:    nil,
			wantErr: errors.New("invalid en passant square: e2"),
		},
		{
			name:    "Invalid halfmove clock should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - x 1",
			want:    nil,
			wantErr: errors.New("strconv.Atoi: parsing \"x\": invalid syntax"),
		},
		{
			name:    "Invalid fullmove number should return an error",
			fen:     "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 x",
			want:    nil,
			wantErr: errors.New("strconv.Atoi: parsing \"x\": invalid syntax"),
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ToPosition(tt.fen)
			if tt.wantErr != nil && (err == nil || err.Error() != tt.wantErr.Error()) {
				t.Errorf("ToPosition() error = %v, wantErr %v", err, tt.wantErr)
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("ToPosition() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func standardPosition() *engine.Position {
	p := engine.NewPosition()
	p.ActiveColor = engine.White
	p.CastlingRights = engine.WhiteKingside | engine.WhiteQueenside | engine.BlackKingside | engine.BlackQueenside
	p.HalfmoveNumber = 2
	p.Put(engine.WhiteRook, engine.A1)
	p.Put(engine.WhiteKnight, engine.B1)
	p.Put(engine.WhiteBishop, engine.C1)
	p.Put(engine.WhiteQueen, engine.D1)
	p.Put(engine.WhiteKing, engine.E1)
	p.Put(engine.WhiteBishop, engine.F1)
	p.Put(engine.WhiteKnight, engine.G1)
	p.Put(engine.WhiteRook, engine.H1)
	p.Put(engine.BlackRook, engine.A8)
	p.Put(engine.BlackKnight, engine.B8)
	p.Put(engine.BlackBishop, engine.C8)
	p.Put(engine.BlackQueen, engine.D8)
	p.Put(engine.BlackKing, engine.E8)
	p.Put(engine.BlackBishop, engine.F8)
	p.Put(engine.BlackKnight, engine.G8)
	p.Put(engine.BlackRook, engine.H8)
	for _, f := range engine.Files {
		p.Put(engine.WhitePawn, engine.SquareOf(f, engine.Rank2))
		p.Put(engine.BlackPawn, engine.SquareOf(f, engine.Rank7))
	}
	return p
}
